third_party/bigframes_vendored/ibis/backends/sql/compilers/bigquery/__init__.py

```diff
@@ -0,0 +1,1106 @@
+                "Please rename your columns to have fewer characters."
+            )
+        return candidate
+
+    def visit_CountStar(self, op, *, arg, where):
+        if where is not None:
+            return self.f.countif(where)
+        return self.f.count(STAR)
+
+    def visit_CountDistinctStar(self, op, *, where, arg):
+        # Bigquery does not support count(distinct a,b,c) or count(distinct (a, b, c))
+        # as expressions must be "groupable":
+        # https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#group_by_grouping_item
+        #
+        # Instead, convert the entire expression to a string
+        # SELECT COUNT(DISTINCT concat(to_json_string(a), to_json_string(b)))
+        # This works with an array of datatypes which generates a unique string
+        # https://cloud.google.com/bigquery/docs/reference/standard-sql/json_functions#json_encodings
+        row = sge.Concat(
+            expressions=[
+                self.f.to_json_string(sg.column(x, quoted=self.quoted))
+                for x in op.arg.schema.keys()
+            ]
+        )
+        if where is not None:
+            row = self.if_(where, row, NULL)
+        return self.f.count(sge.Distinct(expressions=[row]))
+
+    def visit_Degrees(self, op, *, arg):
+        return self._pudf("degrees", arg)
+
+    def visit_Radians(self, op, *, arg):
+        return self._pudf("radians", arg)
+
+    def visit_CountDistinct(self, op, *, arg, where):
+        if where is not None:
+            arg = self.if_(where, arg, NULL)
+        return self.f.count(sge.Distinct(expressions=[arg]))
+
+    def visit_RandomUUID(self, op, **kwargs):
+        return self.f.generate_uuid()
+
+    def visit_ExtractFile(self, op, *, arg):
+        return self._pudf("cw_url_extract_file", arg)
+
+    def visit_ExtractFragment(self, op, *, arg):
+        return self._pudf("cw_url_extract_fragment", arg)
+
+    def visit_ExtractPath(self, op, *, arg):
+        return self._pudf("cw_url_extract_path", arg)
+
+    def visit_ExtractProtocol(self, op, *, arg):
+        return self._pudf("cw_url_extract_protocol", arg)
+
+    def visit_ExtractQuery(self, op, *, arg, key):
+        if key is not None:
+            return self._pudf("cw_url_extract_parameter", arg, key)
+        else:
+            return self._pudf("cw_url_extract_query", arg)
+
+    def _pudf(self, name, *args):
+        name = sg.table(name, db="persistent_udfs", catalog="bigquery-public-data").sql(
+            self.dialect
+        )
+        return self.f[name](*args)
+
+    def visit_DropColumns(self, op, *, parent, columns_to_drop):
+        quoted = self.quoted
+        excludes = [sg.column(column, quoted=quoted) for column in columns_to_drop]
+        star = sge.Star(**{"except": excludes})
+        table = sg.to_identifier(parent.alias_or_name, quoted=quoted)
+        column = sge.Column(this=star, table=table)
+        return sg.select(column).from_(parent)
+
+    def visit_TableUnnest(
```

Comment:
I think this is the unnest we want (has option for `offset` column). If we don't already, let's make sure to track an update to our `explode` to use this instead of the `generate_array` workaround.

Reply:
Creating b/365616029 for tracking.